File: src/construction_mobile/lib/features/notifications/presentation/notification_deep_link.dart
import 'dart:convert';

import '../../../core/router/app_routes.dart';
import '../data/models/app_notification.dart';

/// Turns the JSON payload the API attaches to a notification into an in-app
/// route, or `null` when the notification has nowhere to go.
///
/// Directory routes are withheld from roles the API would answer 403 to, so a
/// tap never leads to a permission error.
String? deepLinkFor(
  AppNotification notification, {
  required bool canViewDirectory,
}) {
  final raw = notification.dataJson;

  if (raw == null || raw.isEmpty) {
    return null;
  }

  final Map<String, dynamic> data;

  try {
    final decoded = jsonDecode(raw);

    if (decoded is! Map<String, dynamic>) {
      return null;
    }

    data = decoded;
  } on FormatException {
    return null;
  }

  return deepLinkForData(
    notification.type,
    data,
    canViewDirectory: canViewDirectory,
  );
}

/// The same routing rules as [deepLinkFor], for a push notification's own
/// data map — an FCM `RemoteMessage.data` carries the identical keys the
/// stored notification's `dataJson` does, plus `notificationType`
/// (`ProcessOutboxCommand.SendPushAsync` adds it), so a push tap resolves a
/// destination the same way a tap in the in-app inbox does, without waiting
/// for the inbox to be re-fetched first.
String? deepLinkForData(
  String? type,
  Map<String, dynamic> data, {
  required bool canViewDirectory,
}) {
  // What concerns the person's own work needs no directory access, so a
  // worker's tap always lands somewhere useful.
  final own = switch (type) {
    'AccommodationAssigned' => AppRoutes.myHousing,
    'TimeEntryRejected' || 'ShiftAutoClosed' => AppRoutes.timeEntries,
    'AbsenceDecided' || 'AbsenceEditProposed' => AppRoutes.absences,
    'ArticleOrderRequested' || 'ArticleOrderStatusChanged' => AppRoutes.articleOrders,
    'RefundRequested' || 'RefundDecided' => AppRoutes.refunds,
    'WeeklyReportDue' => AppRoutes.weeklyReports,
    'BulletinPosted' => AppRoutes.bulletin,
    'TaskAssigned' || 'DefectAssigned' || 'WorkItemDue' => AppRoutes.workItems,
    _ => null,
  };

  if (own != null) {
    return own;
  }

  if (!canViewDirectory) {
    return null;
  }

  return switch (type) {
    'ProjectAssigned' when data['projectId'] is String =>
      AppRoutes.projectDetail(data['projectId'] as String),
    'EmployeeAssigned' when data['employeeId'] is String =>
      AppRoutes.employeeDetail(data['employeeId'] as String),
    'VehicleAssigned' when data['vehicleId'] is String =>
      AppRoutes.vehicleDetail(data['vehicleId'] as String),
    'ToolAssigned' when data['toolId'] is String =>
      AppRoutes.toolDetail(data['toolId'] as String),
    'MaterialLowStock' when data['materialId'] is String =>
      AppRoutes.materialDetail(data['materialId'] as String),
    'VehicleExpenseRejected' || 'VehicleExpenseSubmitted' => AppRoutes.vehicleExpenses,
    _ => null,
  };
}

File: src/construction_admin/src/features/notifications/notificationDeepLink.ts
import type { Notification, User } from '../../api/types';
import { canAdministerAccounts, canViewDirectory } from '../../auth/authHelpers';
import { paths } from '../../routes/paths';

function parseData(dataJson: string | null): Record<string, string> {
  if (!dataJson) return {};

  try {
    const parsed: unknown = JSON.parse(dataJson);
    return parsed && typeof parsed === 'object' ? (parsed as Record<string, string>) : {};
  } catch {
    return {};
  }
}

/** The local day a notification landed on, for types whose payload has no date of its own. */
function notificationDate(notification: Notification): string {
  return notification.createdAt.slice(0, 10);
}

/**
 * Where a notification's own "open it" click should land — the page for the
 * thing that raised it, not just the inbox it sits in. Mirrors the mobile
 * app's `notification_deep_link.dart` (same `dataJson` payload, same
 * per-type routing), extended to every type the API sends and to this
 * panel's own route catalogue.
 *
 * Returns `null` when there is nowhere to go (a free-typed message has no
 * source entity) or the viewer's role cannot reach that route — a click must
 * never send someone into a page the API would 403 them on.
 */
export function resolveNotificationTarget(
  notification: Notification,
  user: User | null | undefined,
): string | null {
  const data = parseData(notification.dataJson);
  const hasDirectory = canViewDirectory(user);
  const hasAccountAdmin = canAdministerAccounts(user);

  switch (notification.type) {
    case 'ProjectAssigned':
    case 'WeeklyReportDue':
      return hasDirectory && data.projectId ? paths.projectDetail(data.projectId) : null;

    case 'EmployeeAssigned':
      return hasDirectory && data.employeeId ? paths.employeeDetail(data.employeeId) : null;

    case 'VehicleAssigned':
      return hasDirectory && data.vehicleId ? paths.vehicleDetail(data.vehicleId) : null;

    case 'ToolAssigned':
      return hasDirectory && data.toolId ? paths.toolDetail(data.toolId) : null;

    case 'TaskAssigned':
    case 'DefectAssigned':
    case 'DefectReported':
    case 'WorkItemDue':
      return hasDirectory && data.workItemId ? paths.workItemEdit(data.workItemId) : null;

    case 'ShiftAutoClosed':
      return hasDirectory && data.timeEntryId ? paths.timeEntryEdit(data.timeEntryId) : null;

    // No time-entry id in these payloads (the entry is created by the clock-in
    // itself), so the deep link opens the day/project/employee it happened on
    // instead — TimeEntriesListPage expands that project's column and flashes
    // the matching card.
    case 'EmployeeClockedIn':
    case 'EmployeeClockedOut':
    case 'ClockInLocationMismatch':
    case 'UnassignedProjectClockIn': {
      if (!hasDirectory || !data.employeeId) return null;
      const params = new URLSearchParams({
        employeeId: data.employeeId,
        date: notificationDate(notification),
      });
      if (data.projectId) params.set('projectId', data.projectId);
      return `${paths.timeEntries}?${params.toString()}`;
    }

    // The costs page has no per-row highlight yet, so this lands on the list,
    // where the rejected cost carries the "Rejected" status.
    case 'VehicleExpenseRejected':
    case 'VehicleExpenseSubmitted':
      return hasDirectory ? paths.vehicleExpenses : null;

    case 'AccommodationContractExpiring':
      return hasDirectory && data.accommodationId ? paths.accommodationDetail(data.accommodationId) : null;

    case 'MaterialLowStock':
      return hasDirectory && data.materialId ? paths.materialDetail(data.materialId) : null;

    case 'TimeEntryRejected':
      return hasDirectory ? paths.timeEntries : null;

    case 'AbsenceDecided':
      if (!hasDirectory) return null;
      return data.absenceId ? `${paths.absences}?highlight=${data.absenceId}` : paths.absences;

    case 'AbsenceRequested':
    case 'AbsenceEditProposed':
      if (!hasDirectory) return null;
      return data.absenceId ? `${paths.absences}?highlight=${data.absenceId}` : paths.absences;

    case 'BulletinPosted':
      return data.bulletinPostId
        ? `${paths.bulletin}?highlight=${data.bulletinPostId}`
        : paths.bulletin;

    case 'DocumentExpiring':
    case 'DocumentRetentionEnded':
      if (!hasAccountAdmin) return null;
      return data.attachmentId
        ? `${paths.expiringDocuments}?highlight=${data.attachmentId}`
        : paths.expiringDocuments;

    // Free-typed text with no source entity to open.
    case 'GeneralAnnouncement':
    case 'DirectMessage':
    default:
      return null;
  }
}
